### PR TITLE
Ctcp 1467 - Validator should reject messages that contain a non-GB/XI customs office of departure or destination actual

### PR DESCRIPTION
**Pull request by olakunlekuye2**

### CODE REVIEW

File: app/uk/gov/hmrc/transitmovementsvalidator/services/XmlValidationService.scala [hunk 1]
```diff
@@ -142,47 +157,129 @@ class XmlValidationServiceImpl @Inject() (implicit ec: ExecutionContext) extends
...
 
-                  if (!elementValue.equalsIgnoreCase(rootTag)) {
+                  val rootNodeCheck = if (!elementValue.equalsIgnoreCase(rootTag.split(":")(1))) {
                     Either.left(BusinessValidationError("Root node doesn't match with the messageType"))
                   } else {
                     Either.right()
                   }
 
+                  val referenceNumberCheck = if (!referenceNumber.toUpperCase.startsWith("GB") && !referenceNumber.toUpperCase.startsWith("XI")) {
+                    Either.left(BusinessValidationError(s"Invalid reference number: $referenceNumber"))
```
Comment (drnaylor-oc) [inline app/uk/gov/hmrc/transitmovementsvalidator/services/XmlValidationService.scala:272]: We already do this check in the router (the ticket is to bring it to here), so the error message should match that in the router. See:

https://github.com/hmrc/transit-movements-router/blob/93ca96465f3d8d6de2ed6b453744dffecd0b4892/app/uk/gov/hmrc/transitmovementsrouter/services/CustomOfficeExtractorService.scala#L73

Specifically, we also need to ensure we tell the trader which element needs changing.

**Review by drnaylor-oc — COMMENTED** [review]
I'm going to be honest, all this state makes it hard to follow what's going on. That's not really your fault because of what you need to work on top of, but we need to be super clear about what's going on, so please add comments that clearly explain what the various state flags do.

I have a few comments on top of that.

File: app/uk/gov/hmrc/transitmovementsvalidator/services/XmlValidationService.scala (hunk 1 above)
Comment (olakunlekuye2) [reply app/uk/gov/hmrc/transitmovementsvalidator/services/XmlValidationService.scala:272]: Done

File: app/uk/gov/hmrc/transitmovementsvalidator/services/XmlValidationService.scala [hunk 2]
```diff
@@ -132,6 +134,19 @@ class XmlValidationServiceImpl @Inject() (implicit ec: ExecutionContext) extends
           t => parsersByType(t)
         )
 
+      def checkMessageType(rootTag: String): String = {
+        val messageType = MessageType.values.find(_.rootNode.equalsIgnoreCase(rootTag.split(":")(1)))
+
+        messageType match {
```
[the remark this answers is not in the document]
Comment (olakunlekuye2) [reply app/uk/gov/hmrc/transitmovementsvalidator/services/XmlValidationService.scala:140]: Done

File: app/uk/gov/hmrc/transitmovementsvalidator/services/JsonValidationService.scala [hunk 3]
```diff
@@ -125,6 +128,19 @@ class JsonValidationServiceImpl @Inject() extends JsonValidationService {
...
+  def checkMessageType(rootTag: String): String = {
+    val messageType = MessageType.values.find(_.rootNode.equalsIgnoreCase(rootTag.split(":")(1)))
+
+    messageType match {
+      case Some(msgType: DepartureMessageType) if MessageType.departureValues.contains(msgType) =>
+        "OfficeOfDeparture"
+      case Some(msgType: ArrivalMessageType) if MessageType.arrivalValues.contains(msgType) =>
+        "OfficeOfDestinationActual"
+      case _ =>
+        ""
```
[the remark this answers is not in the document]
Comment (olakunlekuye2) [reply app/uk/gov/hmrc/transitmovementsvalidator/services/JsonValidationService.scala:140]: Done
